kbs/docs/initdata.md

```diff
@@ -0,0 +1,237 @@
+    "key2": "value2"
+  }
+}
+```
+
+##### TOML version
+
+If you want to avoid escape characters and use the plaintext of ascii file contents as initdata,
+TOML format will be better. A TOML version initdata looks like the following
+```toml
+algorithm = "sha384"
+version = "0.1.0"
+# digest = "e4744fddbb00dd4201326d16ee5a647debf86562a77f1aca176d23017c1cf88821af5c05c03b1ac9b46afa93b9a1d368"
+
+[data]
+"attestation-agent.json" = '''
+{
+"kbs_addr": "http://172.18.0.1:8080"
+}
+'''
+
+"policy.rego" = '''
+package agent_policy
+
+import future.keywords.in
+import future.keywords.every
+
+import input
+
+# Default values, returned by OPA when rules cannot be evaluated to true.
+default CopyFileRequest := false
+default CreateContainerRequest := false
+default CreateSandboxRequest := true
+default DestroySandboxRequest := true
+default ExecProcessRequest := false
+default GetOOMEventRequest := true
+default GuestDetailsRequest := true
+default OnlineCPUMemRequest := true
+default PullImageRequest := true
+default ReadStreamRequest := false
+default RemoveContainerRequest := true
+default RemoveStaleVirtiofsShareMountsRequest := true
+default SignalProcessRequest := true
+default StartContainerRequest := true
+default StatsContainerRequest := true
+default TtyWinResizeRequest := true
+default UpdateEphemeralMountsRequest := true
+default UpdateInterfaceRequest := true
+default UpdateRoutesRequest := true
+default WaitProcessRequest := true
+default WriteStreamRequest := false'''
+```
+
+### Integrity Binding for Different TEEs
+
+There are multiple ways to binding the integrity of initdata data to the TEE evidence.
+Many TEE platforms supports TEE initdata field. The TEE initdata field could be set by
+the untrusted host when launching the TEE, and the field will be included in the
+TEE evidence for remote attestation.
+
+Platforms and corresponding field of the evidence
+- Intel TDX: `mr_config_id`, 48 bytes. Actually `mr_owner` and `mr_owner_config` have similiar
+attributes, but we select only `mr_config_id` for such use.
+- AMD SNP: `hostdata`, 32 bytes.
+- Arm CCA: `CCA_REALM_PERSONALIZATION_VALUE`, 64 bytes.
+- Intel SGX: `CONFIGID`, 64 bytes.
+
+When users want to deploy a TEE, they need to prepare an initdata metadata. The host
+(probably untrusted) SHOULD start TEE instance with initdata digest as TEE initdata.
+
+The software outside the TEE would deliver the initdata into the TEE in
+some way due to concrete architecture design. This spec does not define the concrete
+way to deliver initdata, but strongly recommend that before using the received
+initdata, software stack inside the TEE **MUST** check whether the digest (calculate
+again) of the received initdata is aligned with the one inside evidence.
+
+Other platforms, such as (v)TPM based platforms, can record the initdata digest
+by extending the PCR before using it. This way will also accomplish the integrity binding
+to the TEE evidence.
+
+### Data Canonicalization Algorithm
```

Comment:
Overall I think separate files (metadata file + one or more data files) could help modification work easier. However this involves
1. We should inject not only one file into the TEE but more than one
2. We need to resort the files by traversing the whole directory and calculate the digest

Compared with this, a single JSON/TOML file will not have the drawbacks. There are trade-offs that I'd like to hear more opinions.

> I realize that the current text tries to accommodate arbitrary data, arbitrary spacing, etc. I don't think the additional complexity of the canonicalization and/or transformation is worth it - at least for the initial version of CoCo. I think it's reasonable to ask users to provide the same "measured data" file as input if they want to produce the same "measured data" digest value - rather than CoCo reordering and/or transforming their Init Data at run time.

I hope that the current version could cover more scenarios not only CoCo, just like [the KBS protocol](https://github.com/confidential-containers/trustee/blob/main/kbs/docs/kbs_attestation_protocol.md).  Although we are currently publishing the proposal in the CoCo community, it itself has wider application scenario value.

Reply:
yes